server/services/v1/workers/worker.go

```diff
@@ -0,0 +1,408 @@
+	hearbeatChan chan<- uint
+}
+
+type Event struct {
+	Success  bool
+	Item     metadata.QueueItem
+	WorkerId uint
+}
+
+func newEvent(success bool, item metadata.QueueItem, workerId uint) Event {
+	return Event{
+		success,
+		item,
+		workerId,
+	}
+}
+
+func newWorker(id uint, queue *metadata.QueueSubspace, txMgr *transaction.Manager, tenantMgr *metadata.TenantManager, sleepTime time.Duration, itemEvent chan<- Event, heartbeatChan chan<- uint) *Worker {
+	return &Worker{
+		id:           id,
+		queue:        queue,
+		txMgr:        txMgr,
+		tenantMgr:    tenantMgr,
+		errCount:     0,
+		done:         make(chan struct{}, 1),
+		sleepTime:    sleepTime,
+		itemEvent:    itemEvent,
+		hearbeatChan: heartbeatChan,
+	}
+}
+
+// Add a slightly random jitter for each worker
+// This helps avoid a situation where every worker
+// starts or peaks at the exact time and all tries to read from the job queue.
+func (w *Worker) jitterSleep() {
+	//nolint:gosec // it is ok to use a weak random generator here
+	jitterSleep := rand.Intn(500)
+	time.Sleep(time.Duration(jitterSleep)*time.Millisecond + w.sleepTime)
+}
+
+func (w *Worker) Start() {
+	w.jitterSleep()
+	w.Loop()
+}
+
+func (w *Worker) Loop() {
+	hbTicker := time.NewTicker(1500 * time.Millisecond)
+	for {
+		select {
+		case <-w.done:
+			log.Info().Msgf("Worker %d shutting down", w.id)
+			return
+		case <-hbTicker.C:
+			w.hearbeatChan <- w.id
+		default:
+			err := w.peekAndProcess()
+			if err != nil {
+				w.errCount += 1
+				// if the worker is getting to many errors in a row
+				// shut the worker down and restart a new one
+				if w.errCount >= MAX_ERROR_COUNT {
+					w.Stop()
+				}
+				return
+			}
+			w.errCount = 0
+		}
+		w.jitterSleep()
+	}
+}
+
+func (w *Worker) Stop() {
+	w.done <- struct{}{}
+}
+
+func (w *Worker) peekAndProcess() error {
+	ctx := context.Background()
+	tx, err := w.txMgr.StartTx(ctx)
+	if err != nil {
+		return err
+	}
+	items, err := w.queue.Peek(ctx, tx, 10)
+	if err != nil {
+		return err
+	}
+	if len(items) == 0 {
+		return tx.Rollback(ctx)
+	}
+	selectedItem, err := w.queue.ObtainLease(ctx, tx, &items[0], LEASE_TIME)
+	log.Debug().Msgf("Worker %d: proceeding it %s", w.id, selectedItem.Id)
+	if err != nil {
+		return err
+	}
+	if err = tx.Commit(ctx); err != nil {
+		return err
+	}
+
+	if err = w.processItem(selectedItem); err != nil {
+		log.Err(err).Msgf("Worker %d: failed to proceeds %s", w.id, selectedItem.Id)
+		return w.handleFailedProcessing(ctx, selectedItem)
+	}
+
+	log.Info().Msgf("Worker %d: Completed %s", w.id, selectedItem.Id)
+	w.itemEvent <- newEvent(true, *selectedItem, w.id)
+	return nil
+}
+
+func (w *Worker) handleFailedProcessing(ctx context.Context, selectedItem *metadata.QueueItem) error {
+	selectedItem.ErrorCount += 1
+	tx, err := w.txMgr.StartTx(ctx)
+	if err != nil {
+		return err
+	}
+	// The queue item has had to many errors the job will remove it from the queue
+	if selectedItem.ErrorCount >= MAX_ERROR_COUNT {
```

Comment:
Yes. If it errors out so many times we will loose the task. At this point I think that is ok. We will log it to keep track. A lost task could be to build an index on a database that is deleted

Reply:
Does it mean user's build-index request will be lost, and he can re-trigger it?

Reply:
It could be lost but we have the `buildCollectionIndex` endpoint that would help us recover from this point. 
A failed build like this is an exceptional case